homeassistant/components/sensor/homematicip_cloud.py

```diff
@@ -0,0 +1,258 @@
+"""
+Support for HomematicIP sensors.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/homematicip/
+"""
+
+import logging
+
+from homeassistant.core import callback
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.dispatcher import dispatcher_connect
+from homeassistant.components.homematicip_cloud import (
+    HomematicipGenericDevice, DOMAIN, EVENT_HOME_CHANGED,
+    ATTR_HOME_LABEL, ATTR_HOME_ID, ATTR_LOW_BATTERY, ATTR_RSSI)
+from homeassistant.const import TEMP_CELSIUS, STATE_OK
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['homematicip_cloud']
+
+ATTR_VALVE_STATE = 'valve_state'
+ATTR_VALVE_POSITION = 'valve_position'
+ATTR_TEMPERATURE_OFFSET = 'temperature_offset'
+
+HMIP_UPTODATE = 'up_to_date'
+HMIP_VALVE_DONE = 'adaption_done'
+HMIP_SABOTAGE = 'sabotage'
+
+STATE_LOW_BATTERY = 'low_battery'
+STATE_SABOTAGE = 'sabotage'
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the HomematicIP sensors devices."""
+    # pylint: disable=import-error, no-name-in-module
+    from homematicip.device import (
+        HeatingThermostat, TemperatureHumiditySensorWithoutDisplay,
+        TemperatureHumiditySensorDisplay)
+
+    _LOGGER.info('Setting up HomeMaticIP accespoint & generic devices')
+    homeid = discovery_info['homeid']
+    home = hass.data[DOMAIN][homeid]
+    devices = [HomematicipAccesspoint(hass, home)]
+    if home.devices is None:
+        return
+    for device in home.devices:
+        devices.append(HomematicipDeviceStatus(hass, home, device))
+        if isinstance(device, HeatingThermostat):
+            devices.append(HomematicipHeatingThermostat(hass, home, device))
+        if isinstance(device, TemperatureHumiditySensorWithoutDisplay):
+            devices.append(HomematicipSensorThermometer(hass, home, device))
+            devices.append(HomematicipSensorHumidity(hass, home, device))
+        if isinstance(device, TemperatureHumiditySensorDisplay):
+            devices.append(HomematicipSensorThermometer(hass, home, device))
+            devices.append(HomematicipSensorHumidity(hass, home, device))
+    add_devices(devices)
+
+
+class HomematicipAccesspoint(Entity):
+    """Representation of an HomeMaticIP access point."""
+
+    def __init__(self, hass, home):
+        """Initialize the access point sensor."""
+        self.hass = hass
+        self._home = home
+        dispatcher_connect(
```

Comment:
Move this to the coroutine method `async_added_to_hass`, to avoid a race where the callback is called before the entity has been added to home assistant.